Mongodb dev

delvecja opened the pull request:
<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [ ] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

dcamper commented on plugins/mongodb/README.md:
```diff
@@ -0,0 +1,202 @@
+
+## Installation
+
+The plugin uses vcpkg and can be installed by creating a seperate build directory from the platform and running the following commands:
```
Minor: "separate" instead of "seperate"

dcamper commented on cmake_modules/commonSetup.cmake:
```diff
@@ -165,8 +165,8 @@ IF ("${COMMONSETUP_DONE}" STREQUAL "")
-    ECLBLAS
-    EXAMPLEPLUGIN)
+    EXAMPLEPLUGIN
+    MONGODBEMBED)
```
Minor: Please shift MONGODBEMBED up one line, so EXAMPLEPLUGIN is the last item.

delvecja replied:
I don't remember removing ECLBLAS. I'll add that back as well.

dcamper commented on plugins/mongodb/README.md:
```diff
@@ -0,0 +1,202 @@
+
+## Usage
+
+To start using the plugin you must import mongodb before you use it in the EMBED statement.
```
May want to make this clear that you're talking about ECL code, here.

dcamper commented on plugins/mongodb/README.md:
````diff
@@ -0,0 +1,202 @@
+
+To start using the plugin you must import mongodb before you use it in the EMBED statement.
+```
+Import MongoDB;
````
Style:  Capitalize IMPORT

dcamper commented on plugins/mongodb/README.md:
````diff
@@ -0,0 +1,202 @@
+Import MongoDB;
+```
+
+You then need to declare a function that embeds the MongoDB plugin. Within the EMBED statement is where you can write code to interact with the MongoDB database or collection that you have passed into the EMBED statement. The plugin only supports one line and one operation per EMBED statement. If you wish to chain multiple operations together the aggregate function is supported and can take an ag ...
````
"function that embeds the MongoDB" should be "function that interfaces with MongoDB" or something, right?  You're really creating a function at this point.

dcamper commented on plugins/mongodb/README.md:
```diff
@@ -0,0 +1,202 @@
+
+You then need to declare a function that embeds the MongoDB plugin. Within the EMBED statement is where you can write code to interact with the MongoDB database or collection that you have passed into the EMBED statement. The plugin only supports one line and one operation per EMBED statement. If you wish to chain multiple operations together the aggregate function is supported and can take an ag ...
+
+To create the uri for the MongoDB connection instance the ECL user needs to pass in the username, password, server name, or just the port to use for connecting to the cluster. The plugin takes all of these and creates a shared connection instance for many threads to have access to the MongoDB databases. The plugin can only connect to one MongoDB cluster per workunit, but it can access multiple da ...
```
Suggestion:  Add a list of extra arguments (user, password, server, etc) and what they mean, along with example values if appropriate.

dcamper commented on plugins/mongodb/README.md:
```diff
@@ -0,0 +1,202 @@
+
+### Limitations
+
+There are a few limitations placed by MongoDB on the amount of data you are allowed to transer if you are using a shared cluster. In a single 7-day rolling period you are allowed to transfer:
```
Is this limitation specific to a free tier?

delvecja replied:
It is limited to the shared clusters. The M0 is free, but the M2 and M5 are paid. There is a dedicated tier that is not limited.

dcamper commented on plugins/mongodb/README.md:
```diff
@@ -0,0 +1,202 @@
+
+### Inserting Documents
+
+The plugin supports inserting documents one at a time using `insert_one` and inserting multiple documents at once using insert_many. To insert a document you just need to call insert() inside the embedded script and the plugin will call `insert_one` if you pass in a single record or scalar values. If you pass in a dataset the plugin will automatically call `insert_many` for you.
```
Before jumping in to examples, it might make sense to explain the $ convention.

dcamper commented on plugins/mongodb/README.md:
````diff
@@ -0,0 +1,202 @@
+The plugin can update one or many documents which is specified by the user. it returns a single Record with two fields holding the number of documents that were matched by the filter and another holding the number of documents that were modified by the operation. The update methods can take either two documents or a document and a pipeline. A pipeline follows the structure [{doc}, ... ].
+
+```
+dataset({Integer matched_count, Integer modified_count}) updateOne(REAL raise) := EMBED(mongodb : user(user), password(pwd), server(server), database(db),  collection(coll))
````
Is the RECORD definition returned by MongoDB update functions standardized?  If so, it may make sense to create an exported definition within your .ecllib file so that ECL users can simply use it rather than redefine it themselves every time (and get it wrong some of the time).

dcamper commented on plugins/mongodb/README.md:
````diff
@@ -0,0 +1,202 @@
+
+
+```
+dataset({Integer deleted_count}) deleteOne(REAL min) := EMBED(mongodb : user(user), password(pwd), server(server), database(db),  collection(coll))
````
Also a standardized RECORD definition, perhaps?

delvecja replied:
This can be standard RECORD definition.

dcamper commented on plugins/mongodb/README.md:
````diff
@@ -0,0 +1,202 @@
+```
+
+In this example an index is created with the keys first and last and in ascending order.
+
````
With each of the MongoDB examples above:  Does it make sense to link to MongoDB's documentation at any point?

delvecja commented on plugins/mongodb/README.md:
````diff
@@ -0,0 +1,202 @@
+
+
+```
+dataset({STRING first, STRING last, REAL salary}) aggregate(REAl max, REAL min, INTEGER order) := EMBED(mongodb : user(user), password(pwd), server(server), database(db),  collection(coll), batchSize(100))
````
This one is not standardized.

dcamper commented on plugins/mongodb/examples/mongodb-test.ecl:
```diff
@@ -0,0 +1,188 @@
+layoutUpdate := {Integer matched_count, Integer modified_count};
+layoutperson := {String username, String address, String email};
+
+// Example/Test functions
```
Please add a one-line comment to each example, roughly describing what each is supposed to be doing.

dcamper commented on plugins/mongodb/examples/mongodb-test.ecl:
```diff
@@ -0,0 +1,188 @@
+    aggregate([{$match: { price: { $gte: $min, $lt: $max}}}, {$sort: {price: $asc, extra_people: $asc, security_deposit: $asc}}]);
+ENDEMBED; 
+
+// dataset({INTEGER inserted_count}) insertMany(dataset(layoutEmployee) employees) := Embed(mongodb : user(user), password(pwd), server(server), database('mydb'),  collection('test2'))
```
Remove commented-out stuff.  Or comment it back in and use it.

dcamper commented on plugins/mongodb/examples/mongodb-test.ecl:
```diff
@@ -0,0 +1,188 @@
+    OUTPUT('Done', Named('Status'));
+);
+
+// $or is not allowed in the M0 tier of MongoDB atlas
```
Commented-out code is useful or no?

delvecja replied:
I thought it would be useful to have more examples, but for functions that made changes to the database I had those commented out just so I wasn't constantly updating the database. The $or function I thought would be good to show how complex it can get, but I can't run it on my tier so I commented it out.

dcamper commented on plugins/mongodb/mongodbembed.cpp:
```diff
@@ -0,0 +1,2260 @@
+static const char *g_moduleDescription = "MongoDB Embed Helper";
+static const char *g_version = "MongoDB Embed Helper 1.0.0";
+static const char *g_compatibleVersions[] = { g_version, nullptr };
+static const NullFieldProcessor NULLFIELD(NULL);
```
Style: This and earlier static variable names should be UPPER_CAMEL_CASE and without a special prefix (e.g. g_).

dcamper commented on plugins/mongodb/mongodbembed.cpp:
```diff
@@ -0,0 +1,2260 @@
+
+        while(*start != '}' || depth > 1) 
+        {
+            if(*start == '{') // If we see a document increase the depth
```
Here and everywhere, please insert a space between 'if' and the parenthesis.

dcamper commented on plugins/mongodb/mongodbembed.cpp:
```diff
@@ -0,0 +1,2260 @@
+        int depth = 0; // Keeps track of depth so we don't exit on wrong '}'
+        const char * lastBrkt;
+
+        while(*start != '}' || depth > 1) 
```
Space after 'while' as well.  Basically, all keywords like that.

dcamper commented on plugins/mongodb/mongodbembed.cpp:
```diff
@@ -0,0 +1,2260 @@
+     * @param field RtlFielInfo holds a lot of information about the embed context.
+     * @param query Object holding the bsoncxx::builder::basic::document.
+     */
+    void bindUIntParam(unsigned __int64 value, const RtlFieldInfo * field, std::shared_ptr<MongoDBQuery> query)
```
Inserting thought here:  Does your README talking about the int vs uint issue?

delvecja replied:
No it does not. There should be a section about type conversion between ECL and MongoDB because there are a few non trivial cases.

dcamper commented:
@delvecja Back to you for the items I've identified so far.  I did not finish your .cpp file; I will in the next round.

Looks very good so far!

delvecja commented:
@dcamper I believe I have resolved everything you brought up. I added a standard return type for the update function to the ecllib file and changed the return type of the delete method to an integer. I added information about the datatypes, parameter passing, and options in the readme. Back to you.

dcamper commented on plugins/mongodb/README.md:
```diff
@@ -0,0 +1,248 @@
+
+### Parameters
+
+To pass paremeters of the function you are defining into the MongoDB query insert them anywhere in the BSON document with a leading '$'.
```
Nit-pick:  I think the wording here is "to *use* function parameters within the MongoDB statement, prefix them with a '$'" or something along those lines.  You're passing them just fine now, without the $.

dcamper commented on plugins/mongodb/mongodb.ecllib:
```diff
@@ -0,0 +1,26 @@
+EXPORT boolean supportsImport := false;
+EXPORT boolean supportsScript := true;
+
+EXPORT updateRecord := {INTEGER matched_count, INTEGER modified_count};
```
Recommend renaming 'updateRecord' for clarity (it currently sounds like a verb).  This is, in this case, a result of call; maybe use 'updateRecordResult' ?

dcamper commented on plugins/mongodb/mongodbembed.hpp:
```diff
@@ -0,0 +1,736 @@
+#include "roxiemem.hpp"
+#include <time.h>
+
+#include <vector>
```
Comment:  We normally try to minimize the included headers by shifting everything possible to the .cpp file.  If you have time, it may be worthwhile to go through that exercise.

dcamper commented:
@delvecja Back to you.  Just a couple of minor comments.  Once you are finished (and everything passes tests), and you thing you will have no further commits, please collapse all of your commits to one commit ("collapse" or "rebase" -- your choice of terminology).  The final goal is to have a single commit to merge.

jackdelv committed:
Added MongoDB Plugin
